### PR TITLE
Add ruby memory profiler option

### PR DESCRIPTION
**Pull request by busterb**
This adds a new way to profile Metasploit for memory usage, pulled out of #12561 

To get memory profiler results, invoke msfconsole with something like this: `METASPLOIT_MEMORY_PROFILE=1 ./msfconsole`. This will enable the memory profiler and write the results to a file called `memory.profile`.

There are other enhancements to be made in how we measure utilization that I'll explore later, this is up for now as a way to separate the profiler changes from the changes to actual memory utilization in the previous PR.

### CODE REVIEW

**Comment by ccondon-r7** [comment]
@adfoster-r7 Sounds like this is probably still a useful idea, even if you think a different implementation would be better. Thoughts?

**Comment by adfoster-r7** [comment]
This looks pretty good to me, i'd been using rbspy - but this captures start up performance costs better IMO :+1:

File: lib/metasploit/framework/command/console.rb [hunk 1]
```diff
@@ -39,12 +39,50 @@ def spinner
...
+        end
+      }
+    end
+
+    if ENV['METASPLOIT_MEMORY_PROFILE']
+      require 'memory_profiler'
+      MemoryProfiler.start
+      at_exit {
+        report = MemoryProfiler.stop
+        report.pretty_print(to_file: 'memory.profile')
```
Comment (adfoster-r7) [inline lib/metasploit/framework/command/console.rb]: I'll move this into `tmp` and add logging for consistency with the above approach :+1:

File: lib/metasploit/framework/command/console.rb [hunk 2]
```diff
@@ -39,12 +39,50 @@ def spinner
     end
   end
 
+  def start_profiler
+    if ENV['METASPLOIT_CPU_PROFILE']
+      gem 'perftools.rb'
```
Comment (adfoster-r7) [inline lib/metasploit/framework/command/console.rb]: Do we want to add perftools.rb to the Gemfile?

```
perftools.rb is not part of the bundle. Add it to your Gemfile.local
```

It doesn't seem to compile for mac + ruby 2.6.5 though 🤔

Comment (adfoster-r7) [reply lib/metasploit/framework/command/console.rb]: After poking further, this issue might have existed for a few years - https://github.com/tmm1/perftools.rb/issues/78 and the project might be abandoned?

I'll try swapping this out for stackprof 👀

File: lib/metasploit/framework/command/console.rb [hunk 3]
```diff
@@ -39,12 +39,50 @@ def spinner
...
+      require 'perftools'
+
+      formatted_time = Time.now.strftime('%Y%m%d%H%M%S')
+      root = Pathname.new(__FILE__).parent
```
Comment (adfoster-r7) [inline lib/metasploit/framework/command/console.rb]: I think moving this logic from the root directory to inside `commands` now calculates the wrong root directory, will fix 🤞

**Comment by adfoster-r7** [comment]
Thanks for the awesome foundation work, I'm going ahead and pulling this work into #13057 to get this work delivered 🚢

**Pull request closed without merging**